File: ledger_reconciler/ledger_profiles.py
"""Column-name profiles for the ledger exports of the accounting platforms
most small US/UK/AU clients actually use, so onboarding a new client on a
platform we already support is a config lookup, not a rewrite.

Sourced against real documented export formats where confirmed:
- Xero "Account Transactions" export: Account Code, Credit, Debit,
  Description, Reference, Source.
- Wave transaction export: Date, Description, Amount.
- FreshBooks expense export: Date, Vendor, Category, Amount, Currency, Tax.
- QuickBooks Online "Transaction List by Date": the well-established common
  column set (Date, Transaction Type, Num, Name, Memo/Description, Account,
  Amount) — NOT independently confirmed against a live export the way the
  other three were. Treat this one profile as lower-confidence; if a real
  QuickBooks export doesn't match, it's a one-line alias fix here, not code.

Each profile lists ALIASES per logical field (matched case-insensitively
against the header row), same idea as the multiple label patterns in
extract.py — a platform's exact wording varies by region/version, so we
match on a short list of known variants rather than one exact string.
"""
from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class LedgerProfile:
    name: str
    date_aliases: tuple[str, ...]
    description_aliases: tuple[str, ...]
    amount_aliases: tuple[str, ...] = ()
    debit_aliases: tuple[str, ...] = ()   # used instead of amount_aliases when
    credit_aliases: tuple[str, ...] = ()  # the platform splits debit/credit
    reference_aliases: tuple[str, ...] = ()

    @property
    def uses_debit_credit_split(self) -> bool:
        return bool(self.debit_aliases or self.credit_aliases)


PROFILES: list[LedgerProfile] = [
    LedgerProfile(
        name="Xero",
        date_aliases=("date",),
        description_aliases=("description",),
        debit_aliases=("debit",),
        credit_aliases=("credit",),
        reference_aliases=("reference",),
    ),
    LedgerProfile(
        name="Wave",
        date_aliases=("date",),
        description_aliases=("description",),
        amount_aliases=("amount",),
    ),
    LedgerProfile(
        name="FreshBooks",
        date_aliases=("date",),
        description_aliases=("vendor", "category"),
        amount_aliases=("amount",),
    ),
    LedgerProfile(
        name="QuickBooks",  # lower confidence — see module docstring
        date_aliases=("date",),
        description_aliases=("memo/description", "memo", "description", "name"),
        amount_aliases=("amount",),
        reference_aliases=("num",),
    ),
    LedgerProfile(
        name="Generic",  # the plain Date/Description/Reference/Amount shape
        date_aliases=("date",),
        description_aliases=("description",),
        amount_aliases=("amount",),
        reference_aliases=("reference",),
    ),
]


def _first_match(headers: list[str], aliases: tuple[str, ...]) -> int | None:
    for alias in aliases:
        if alias in headers:
            return headers.index(alias)
    return None


def _try_profile(profile: LedgerProfile, headers: list[str]) -> dict[str, int] | None:
    date_i = _first_match(headers, profile.date_aliases)
    desc_i = _first_match(headers, profile.description_aliases)
    if date_i is None or desc_i is None:
        return None

    cols = {"date": date_i, "description": desc_i}

    if profile.uses_debit_credit_split:
        debit_i = _first_match(headers, profile.debit_aliases)
        credit_i = _first_match(headers, profile.credit_aliases)
        if debit_i is None and credit_i is None:
            return None
        if debit_i is not None:
            cols["debit"] = debit_i
        if credit_i is not None:
            cols["credit"] = credit_i
    else:
        amount_i = _first_match(headers, profile.amount_aliases)
        if amount_i is None:
            return None
        cols["amount"] = amount_i

    ref_i = _first_match(headers, profile.reference_aliases)
    if ref_i is not None:
        cols["reference"] = ref_i

    return cols


def detect_profile(headers: list[str]) -> tuple[LedgerProfile, dict[str, int]] | None:
    """Tries every profile and returns whichever one resolves the MOST
    columns, not just the first whose bare minimum is satisfied. A generic
    Date/Description/Reference/Amount sheet also happens to satisfy Wave's
    (looser) requirements, for example — picking by first-match-wins would
    silently drop the reference column that was actually there. ``headers``
    must already be lowercased/stripped."""
    best: tuple[LedgerProfile, dict[str, int]] | None = None
    for profile in PROFILES:
        cols = _try_profile(profile, headers)
        if cols is None:
            continue
        if best is None or len(cols) > len(best[1]):
            best = (profile, cols)
    return best

File: web/src/domain/ledgerProfiles.ts
/**
 * Column-name profiles for the ledger exports of the accounting platforms
 * most small US/UK/AU clients actually use. Mirrors
 * ledger_reconciler/ledger_profiles.py.
 *
 * Adding support for a new platform means adding one entry here, not
 * touching the reconciliation logic. Each field lists ALIASES, matched
 * case-insensitively against the header row, because a platform's exact
 * wording varies by region and version.
 *
 * Sourced against documented export formats where confirmed. The
 * QuickBooks entry is the well-known common column set but was not
 * verified against a live export the way the other three were, so it is
 * deliberately marked lower-confidence: if a real export disagrees, the
 * fix is one alias here, not a code change.
 */

export interface LedgerProfile {
  name: string
  dateAliases: string[]
  descriptionAliases: string[]
  amountAliases?: string[]
  /** Used instead of amountAliases when the platform splits debit/credit. */
  debitAliases?: string[]
  creditAliases?: string[]
  referenceAliases?: string[]
}

export function usesDebitCreditSplit(p: LedgerProfile): boolean {
  return Boolean(p.debitAliases?.length || p.creditAliases?.length)
}

export const PROFILES: LedgerProfile[] = [
  {
    name: 'Xero',
    dateAliases: ['date'],
    descriptionAliases: ['description'],
    debitAliases: ['debit'],
    creditAliases: ['credit'],
    referenceAliases: ['reference'],
  },
  {
    name: 'Wave',
    dateAliases: ['date'],
    descriptionAliases: ['description'],
    amountAliases: ['amount'],
  },
  {
    name: 'FreshBooks',
    dateAliases: ['date'],
    descriptionAliases: ['vendor', 'category'],
    amountAliases: ['amount'],
  },
  {
    name: 'QuickBooks', // lower confidence, see module comment
    dateAliases: ['date'],
    descriptionAliases: ['memo/description', 'memo', 'description', 'name'],
    amountAliases: ['amount'],
    referenceAliases: ['num'],
  },
  {
    name: 'Generic',
    dateAliases: ['date'],
    descriptionAliases: ['description'],
    amountAliases: ['amount'],
    referenceAliases: ['reference'],
  },
]

/** Which header index each logical field resolved to. */
export type ResolvedColumns = Partial<
  Record<'date' | 'description' | 'amount' | 'debit' | 'credit' | 'reference', number>
>

export interface Detection {
  profile: LedgerProfile
  columns: ResolvedColumns
}

function firstMatch(headers: string[], aliases: string[] | undefined): number | null {
  for (const alias of aliases ?? []) {
    const i = headers.indexOf(alias)
    if (i !== -1) return i
  }
  return null
}

function tryProfile(p: LedgerProfile, headers: string[]): ResolvedColumns | null {
  const dateIdx = firstMatch(headers, p.dateAliases)
  const descIdx = firstMatch(headers, p.descriptionAliases)
  if (dateIdx === null || descIdx === null) return null

  const columns: ResolvedColumns = { date: dateIdx, description: descIdx }

  if (usesDebitCreditSplit(p)) {
    const debitIdx = firstMatch(headers, p.debitAliases)
    const creditIdx = firstMatch(headers, p.creditAliases)
    if (debitIdx === null && creditIdx === null) return null
    if (debitIdx !== null) columns.debit = debitIdx
    if (creditIdx !== null) columns.credit = creditIdx
  } else {
    const amountIdx = firstMatch(headers, p.amountAliases)
    if (amountIdx === null) return null
    columns.amount = amountIdx
  }

  const refIdx = firstMatch(headers, p.referenceAliases)
  if (refIdx !== null) columns.reference = refIdx

  return columns
}

/**
 * Returns whichever profile resolves the MOST columns, not simply the first
 * whose minimum is satisfied.
 *
 * This mattered in practice: a generic Date/Description/Reference/Amount
 * sheet also satisfies Wave's looser requirements, and first-match-wins
 * picked Wave and silently dropped the reference column that was sitting
 * right there, degrading exact-reference matching to amount-and-date
 * guessing. `headers` must already be lowercased and trimmed.
 */
export function detectProfile(headers: string[]): Detection | null {
  let best: Detection | null = null
  for (const profile of PROFILES) {
    const columns = tryProfile(profile, headers)
    if (!columns) continue
    if (best === null || Object.keys(columns).length > Object.keys(best.columns).length) {
      best = { profile, columns }
    }
  }
  return best
}
